##### 1. Import numpy as np and see the version
Difficulty Level: L1

    Q. Import numpy as np and print the version number.

In [56]:
import numpy as np

np.__version__

'1.26.4'

##### 2. How to create a 1D array?
Difficulty Level: L1

    Q. Create a 1D array of numbers from 0 to 9

    Desired output:

    #> array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [57]:
a = np.arange(10)

a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

##### 3. How to create a boolean array?
    Difficulty Level: L1

    Q. Create a 3×3 numpy array of all True’s

In [58]:
# Boa pratica definir o tipo dos valores
a = np.full((3,3), True, dtype=bool)

a

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

##### 4. How to extract items that satisfy a given condition from 1D array?
Difficulty Level: L1

    Q. Extract all odd numbers from arr

    Input:

    arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

    Desired output:

    #> array([1, 3, 5, 7, 9])

In [59]:
arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

arr[arr % 2 == 1]

array([1, 3, 5, 7, 9])

##### 5. How to replace items that satisfy a condition with another value in numpy array?
Difficulty Level: L1

    Q. Replace all odd numbers in arr with -1

    Input:

    arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
    Desired Output:

    #>  array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1])

In [60]:
arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

arr[arr % 2 == 1] = -1

arr

array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1])

##### 6. How to replace items that satisfy a condition without affecting the original array?
    Difficulty Level: L2

    Q. Replace all odd numbers in arr with -1 without changing arr

    Input:

    arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
    Desired Output:

    out
    #>  array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1])
    arr
    #>  array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [61]:
arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

# Where basicamente executa uma condição de busca em um array
# Pode também alterar valores no retorno
# Condição, valor a ser trocado e array original
out = np.where(arr % 2 == 1, -1, arr)

print(out)

arr

[ 0 -1  2 -1  4 -1  6 -1  8 -1]


array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

##### 7. How to reshape an array?
    Difficulty Level: L1

    Q. Convert a 1D array to a 2D array with 2 rows

    Input:

    np.arange(10)

    #> array([ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
    Desired Output:

    #> array([[0, 1, 2, 3, 4],
    #>        [5, 6, 7, 8, 9]])

In [62]:
arr = np.arange(10)

arr = arr.reshape((2,5))

arr

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

* Outra solução <br><br>
Definir a coluna como -1 automaticamente seleciona o número de colunas com base no número de linhas, o mesmo é valido para linha com valor -1

In [63]:
arr = np.arange(10)

arr = arr.reshape((2, -1))

arr

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

##### 8. How to stack two arrays vertically?
    Difficulty Level: L2

    Q. Stack arrays a and b vertically

    Input

    a = np.arange(10).reshape(2,-1)
    b = np.repeat(1, 10).reshape(2,-1)
    Desired Output:

    #> array([[0, 1, 2, 3, 4],
    #>        [5, 6, 7, 8, 9],
    #>        [1, 1, 1, 1, 1],
    #>        [1, 1, 1, 1, 1]])

In [64]:
a = np.arange(10).reshape(2, -1)
b = np.repeat(1, 10).reshape(2, -1)

# np.vstack recebe uma lista de arrays como parametro
arr = np.vstack([a,b])

arr

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1]])

##### Difficulty Level: L2

    Q. Stack the arrays a and b horizontally.

    Input

    a = np.arange(10).reshape(2,-1)

    b = np.repeat(1, 10).reshape(2,-1)
    Desired Output:

    #> array([[0, 1, 2, 3, 4, 1, 1, 1, 1, 1],
    #>        [5, 6, 7, 8, 9, 1, 1, 1, 1, 1]])

In [65]:
a = np.arange(10).reshape(2, -1)

b = np.repeat(1, 10).reshape(2, -1)

arr = np.hstack([a,b])

arr

array([[0, 1, 2, 3, 4, 1, 1, 1, 1, 1],
       [5, 6, 7, 8, 9, 1, 1, 1, 1, 1]])

##### 10. How to generate custom sequences in numpy without hardcoding?
    Difficulty Level: L2

    Q. Create the following pattern without hardcoding. Use only numpy functions and the below input array a.

    Input:

    a = np.array([1,2,3])`
    Desired Output:

    #> array([1, 1, 1, 2, 2, 2, 3, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

In [66]:
a = np.array([1,2,3])

np.r_[np.repeat(a, 3), np.tile(a, 3)]

array([1, 1, 1, 2, 2, 2, 3, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

* np.r_ é uma maneira de concatenar arrays em numpy. Nessa caso, np.repeat está repetindo cada elemento de a 3 vezes, enquanto np.tile repete a como um todo 3 vezes

##### 11. How to get the common items between two python numpy arrays?
    Difficulty Level: L2

    Q. Get the common items between a and b

    Input:

    a = np.array([1,2,3,2,3,4,3,4,5,6])
    b = np.array([7,2,10,2,7,4,9,4,9,8])
    Desired Output:

    array([2, 4])


In [67]:
a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])

# A função intersect1d recupera elementos em comum a dois arrays
np.intersect1d(a,b)

array([2, 4])

##### 12. How to remove from one array those items that exist in another?
    Difficulty Level: L2

    Q. From array a remove all items present in array b

    Input:

    a = np.array([1,2,3,4,5])
    b = np.array([5,6,7,8,9])
    Desired Output:

    array([1,2,3,4])

In [68]:
a = np.array([1, 2, 3, 4, 5])
b = np.array([5, 6, 7, 8, 9])

# A função setdiff1d remove todos os elementos de a que estão presentes em b
np.setdiff1d(a,b)

array([1, 2, 3, 4])

##### 13. How to get the positions where elements of two arrays match?
    Difficulty Level: L2

    Q. Get the positions where elements of a and b match

    Input:

    a = np.array([1,2,3,2,3,4,3,4,5,6])
    b = np.array([7,2,10,2,7,4,9,4,9,8])
    Desired Output:

    #> (array([1, 3, 5, 7]),)

In [69]:
a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])

# Novamente, usa-se o where, pois funciona como uma consulta no array
np.where(a == b)

(array([1, 3, 5, 7]),)

##### 14. How to extract all numbers between a given range from a numpy array?
    Difficulty Level: L2

    Q. Get all items between 5 and 10 from a.

    Input:

    a = np.array([2, 6, 1, 9, 10, 3, 27])
    Desired Output:

    (array([6, 9, 10]),)

In [70]:
a = np.array([2, 6, 1, 9, 10, 3, 27])

index = np.where((a >= 5) & (a <= 10))

a[index]

array([ 6,  9, 10])

* Nesse caso, usa-se np.where para recuperar os indices que cumprem essa condicao. Depois aplica-se os indices à a

##### 15. How to make a python function that handles scalars to work on numpy arrays?
    Difficulty Level: L2

    Q. Convert the function maxx that works on two scalars, to work on two arrays.

    Input:

    def maxx(x, y):
        """Get the maximum of two items"""
        if x >= y:
            return x
        else:
            return y

    maxx(1, 5)
    #> 5
    Desired Output:

    a = np.array([5, 7, 9, 8, 6, 4, 5])
    b = np.array([6, 3, 4, 8, 9, 7, 1])
    pair_max(a, b)
    #> array([ 6.,  7.,  9.,  8.,  9.,  7.,  5.])

In [71]:
def maxx(x, y):
    if x >= y:
        return x
    else:
        return y


# np.vectorize transforma funções que aplicam sobre um único valor em funções que podem aplicar sobre arrays
# São passados como parâmetros: a função a ser vetorizada e o tipod e saida da função
# Nesse caso, paix_max agora é uma função que pode lidar com vetores np
pair_max = np.vectorize(maxx, otypes=[float])

a = np.array([5, 7, 9, 8, 6, 4, 5])
b = np.array([6, 3, 4, 8, 9, 7, 1])

pair_max(a, b)

array([6., 7., 9., 8., 9., 7., 5.])

##### Difficulty Level: L2

    Q. Swap columns 1 and 2 in the array arr.

    arr = np.arange(9).reshape(3,3)
    arr

In [72]:
arr = np.arange(9).reshape(3, 3)

arr[:, [0, 2, 1]]

array([[0, 2, 1],
       [3, 5, 4],
       [6, 8, 7]])

##### 17. How to swap two rows in a 2d numpy array?
    Difficulty Level: L2

    Q. Swap rows 1 and 2 in the array arr:

    arr = np.arange(9).reshape(3,3)
    arr

In [73]:
arr = np.arange(9).reshape(3, 3)

arr[[0,2,1],:]

array([[0, 1, 2],
       [6, 7, 8],
       [3, 4, 5]])

##### 18. How to reverse the rows of a 2D array?
    Difficulty Level: L2

    Q. Reverse the rows of a 2D array arr.

    # Input
    arr = np.arange(9).reshape(3,3)

In [74]:
arr = np.arange(9).reshape(3, 3)

# Usando o passo como -1, indo de trás pra frente
arr[::-1,:]

array([[6, 7, 8],
       [3, 4, 5],
       [0, 1, 2]])

##### 19. How to reverse the columns of a 2D array?
    Difficulty Level: L2

    Q. Reverse the columns of a 2D array arr.

    # Input
    arr = np.arange(9).reshape(3,3)

In [75]:
arr = np.arange(9).reshape(3, 3)

arr[:, ::-1]

array([[2, 1, 0],
       [5, 4, 3],
       [8, 7, 6]])

##### 20. How to create a 2D array containing random floats between 5 and 10?
  Difficulty Level: L2

  Q. Create a 2D array of shape 5x3 to contain random decimal numbers between 5 and 10

In [76]:
# np.random.uniform recebe 3 parametros: minimo, maximo e tamanho do vetor, a sua forma
np.random.uniform(5,10,size=(5,3))

array([[9.325499, 9.151389, 7.690807],
       [9.612347, 5.485732, 5.514237],
       [8.507536, 9.452399, 5.797802],
       [6.377863, 8.362458, 5.821516],
       [8.506856, 7.438176, 8.403389]])

##### 21. How to print only 3 decimal places in python numpy array?
    Difficulty Level: L1

    Q. Print or show only 3 decimal places of the numpy array rand_arr.

    Input:

    rand_arr = np.random.random((5,3))

In [77]:
rand_arr = np.random.random((5, 3))

# usando np.set_printoptions, é possível definir as opcções do printo do array
# nesse caso,e stá definido com uma precisão de  casas decimais

np.set_printoptions(precision=3)

rand_arr

array([[0.522, 0.043, 0.224],
       [0.575, 0.12 , 0.5  ],
       [0.138, 0.053, 0.178],
       [0.442, 0.878, 0.949],
       [0.478, 0.461, 0.637]])

##### 22. How to pretty print a numpy array by suppressing the scientific notation (like 1e10)?
    Difficulty Level: L1

    Q. Pretty print rand_arr by suppressing the scientific notation (like 1e10)

    Input:

    # Create the random array
    np.random.seed(100)
    rand_arr = np.random.random([3,3])/1e3
    rand_arr

    #> array([[  5.434049e-04,   2.783694e-04,   4.245176e-04],
    #>        [  8.447761e-04,   4.718856e-06,   1.215691e-04],
    #>        [  6.707491e-04,   8.258528e-04,   1.367066e-04]])
    Desired Output:

    #> array([[ 0.000543,  0.000278,  0.000425],
    #>        [ 0.000845,  0.000005,  0.000122],
    #>        [ 0.000671,  0.000826,  0.000137]])

In [78]:
np.set_printoptions(suppress=False)
np.random.seed(100)
rand_arr = np.random.random([3, 3]) / 1e3

# Novamente é alterado no printoptions, deixando supress como True
np.set_printoptions(suppress=True, precision=6)  # precision is optional
rand_arr

array([[0.000543, 0.000278, 0.000425],
       [0.000845, 0.000005, 0.000122],
       [0.000671, 0.000826, 0.000137]])

##### 23. How to limit the number of items printed in output of numpy array?
    Difficulty Level: L1

    Q. Limit the number of items printed in python numpy array a to a maximum of 6 elements.

    Input:

    a = np.arange(15)
    #> array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])
    Desired Output:

    #> array([ 0,  1,  2, ..., 12, 13, 14])

In [79]:
# Definir em printoptions para no máximo 6 elementos, isso está localizado em treshold
np.set_printoptions(threshold=6)
a = np.arange(15)
a

array([ 0,  1,  2, ..., 12, 13, 14])

##### 24. How to print the full numpy array without truncating
    Difficulty Level: L1

    Q. Print the full numpy array a without truncating.

    Input:

    np.set_printoptions(threshold=6)
    a = np.arange(15)
    a
    #> array([ 0,  1,  2, ..., 12, 13, 14])
    Desired Output:

    a
    #> array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [80]:
import sys

np.set_printoptions(threshold=6)
a = np.arange(15)

# Para mostrar todos os elementos, é necessário definir treshold como sys.maxsize
np.set_printoptions(threshold= sys.maxsize)
a

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

##### 25. How to import a dataset with numbers and texts keeping the text intact in python numpy?
    Difficulty Level: L2

    Q. Import the iris dataset keeping the text intact.

In [81]:
# armazena a url do dataset
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# obtem os dados a partir da url, definindo como separador a virgula , que separa as amostras no dataset
# É preciso setar o tipo como objeto para poder recuperar o texto, caso contrário ele só recupera os números
iris = np.genfromtxt(url, delimiter=",", dtype="object")
names = ("sepallength", "sepalwidth", "petallength", "petalwidth", "species")


iris[:5, :]

array([[b'5.1', b'3.5', b'1.4', b'0.2', b'Iris-setosa'],
       [b'4.9', b'3.0', b'1.4', b'0.2', b'Iris-setosa'],
       [b'4.7', b'3.2', b'1.3', b'0.2', b'Iris-setosa'],
       [b'4.6', b'3.1', b'1.5', b'0.2', b'Iris-setosa'],
       [b'5.0', b'3.6', b'1.4', b'0.2', b'Iris-setosa']], dtype=object)

##### 26. How to extract a particular column from 1D array of tuples?
    Difficulty Level: L2

    Q. Extract the text column species from the 1D iris imported in previous question

In [82]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_1d = np.genfromtxt(url, delimiter=",", dtype=None)

# itera sobre cada linha do array iris_1d
# recuperando a quarta coluna de cada linha
species = np.array([row[4] for row in iris_1d])

species[0:5]

/tmp/ipykernel_27882/2986695082.py:2: VisibleDeprecationWarning: Reading unicode strings without specifying the encoding argument is deprecated. Set the encoding, use None for the system default.
  iris_1d = np.genfromtxt(url, delimiter=",", dtype=None)


array([b'Iris-setosa', b'Iris-setosa', b'Iris-setosa', b'Iris-setosa',
       b'Iris-setosa'], dtype='|S15')

##### 27. How to convert a 1d array of tuples to a 2d numpy array?
    Difficulty Level: L2

    Q. Convert the 1D iris to 2D array iris_2d by omitting the species text field.

    Input:

    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    iris_1d = np.genfromtxt(url, delimiter=',', dtype=None)

In [83]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# Nesse caso, recupera os dados como sendo do tipo float
# Por meio do parametro usecols ,somente utiliza as 4 primeiras colunas, excluindo a coluna de nome
iris_1d = np.genfromtxt(url, delimiter=",", dtype='float', usecols=[0,1,2,3])

iris_1d[0:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

##### 28. How to compute the mean, median, standard deviation of a numpy array?
    Difficulty: L1

    Q. Find the mean, median, standard deviation of iris's sepallength (1st column)

    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    iris = np.genfromtxt(url, delimiter=',', dtype='object')

In [84]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# Define os dados como float e usa apenas a primeira coluna, coluna desejada
sepallength = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0])


print(f'Mean: {np.mean(sepallength):.2f}')
print(f'Median: {np.median(sepallength):.2f}')
print(f'Stardard deviation: {np.std(sepallength):.2f}')

Mean: 5.84
Median: 5.80
Stardard deviation: 0.83


##### 29. How to normalize an array so the values range exactly between 0 and 1?
    Difficulty: L2

    Q. Create a normalized form of iris's sepallength whose values range exactly between 0 and 1 so that the minimum has value 0 and maximum has value 1.

    Input:

    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    sepallength = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0])

In [85]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
sepallength = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0])

#Cria uma versão normalizada dos valores
Smax, Smin = sepallength.max(), sepallength.min()
S = (sepallength - Smin) / (Smax - Smin)

S

array([0.222222, 0.166667, 0.111111, 0.083333, 0.194444, 0.305556,
       0.083333, 0.194444, 0.027778, 0.166667, 0.305556, 0.138889,
       0.138889, 0.      , 0.416667, 0.388889, 0.305556, 0.222222,
       0.388889, 0.222222, 0.305556, 0.222222, 0.083333, 0.222222,
       0.138889, 0.194444, 0.194444, 0.25    , 0.25    , 0.111111,
       0.138889, 0.305556, 0.25    , 0.333333, 0.166667, 0.194444,
       0.333333, 0.166667, 0.027778, 0.222222, 0.194444, 0.055556,
       0.027778, 0.194444, 0.222222, 0.138889, 0.222222, 0.083333,
       0.277778, 0.194444, 0.75    , 0.583333, 0.722222, 0.333333,
       0.611111, 0.388889, 0.555556, 0.166667, 0.638889, 0.25    ,
       0.194444, 0.444444, 0.472222, 0.5     , 0.361111, 0.666667,
       0.361111, 0.416667, 0.527778, 0.361111, 0.444444, 0.5     ,
       0.555556, 0.5     , 0.583333, 0.638889, 0.694444, 0.666667,
       0.472222, 0.388889, 0.333333, 0.333333, 0.416667, 0.472222,
       0.305556, 0.472222, 0.666667, 0.555556, 0.361111, 0.333

##### 30. How to compute the softmax score?
    Difficulty Level: L3

    Q. Compute the softmax score of sepallength.

    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    sepallength = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0])

In [86]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris = np.genfromtxt(url, delimiter=",", dtype="object")
sepallength = np.array([float(row[0]) for row in iris])



def softmax(x):
    e_x = np.exp(x - np.max(x))
    return e_x / e_x.sum(axis=0)


softmax(sepallength)

array([0.00222 , 0.001817, 0.001488, 0.001346, 0.002008, 0.002996,
       0.001346, 0.002008, 0.001102, 0.001817, 0.002996, 0.001644,
       0.001644, 0.000997, 0.00447 , 0.004044, 0.002996, 0.00222 ,
       0.004044, 0.00222 , 0.002996, 0.00222 , 0.001346, 0.00222 ,
       0.001644, 0.002008, 0.002008, 0.002453, 0.002453, 0.001488,
       0.001644, 0.002996, 0.002453, 0.003311, 0.001817, 0.002008,
       0.003311, 0.001817, 0.001102, 0.00222 , 0.002008, 0.001218,
       0.001102, 0.002008, 0.00222 , 0.001644, 0.00222 , 0.001346,
       0.002711, 0.002008, 0.01484 , 0.008144, 0.013428, 0.003311,
       0.009001, 0.004044, 0.007369, 0.001817, 0.009947, 0.002453,
       0.002008, 0.00494 , 0.005459, 0.006033, 0.003659, 0.010994,
       0.003659, 0.00447 , 0.006668, 0.003659, 0.00494 , 0.006033,
       0.007369, 0.006033, 0.008144, 0.009947, 0.01215 , 0.010994,
       0.005459, 0.004044, 0.003311, 0.003311, 0.00447 , 0.005459,
       0.002996, 0.005459, 0.010994, 0.007369, 0.003659, 0.003

##### 31. How to find the percentile scores of a numpy array?
    Difficulty Level: L1

    Q. Find the 5th and 95th percentile of iris's sepallength

    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    sepallength = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0])

In [87]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
sepallength = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0])


# Calcula o valor dos percentis passados em q
# np.percentile recebe o array em que será aplicado a operação e os valores de percentil a serem calculados
np.percentile(sepallength, q=[5, 95])

array([4.6  , 7.255])

<small>5% dos valores estão abaixo de 4.6 e 95% dos valores estão abaixo de 7.255</small>

##### 32. How to insert values at random positions in an array?
    Difficulty Level: L2

    Q. Insert np.nan values at 20 random positions in iris_2d dataset

    # Input
    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    iris_2d = np.genfromtxt(url, delimiter=',', dtype='object')

In [88]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_2d = np.genfromtxt(url, delimiter=",", dtype="object")

# Define uma semente para gerador de números aleatórios
np.random.seed(100)

# Primeiro gera um array com 20 números aleatórios entre 0 e 149, representando os índices das linhas do array
# Depois gera um array com  20 números aleatórios de 0 até 3, representando as colunas do array
# Com os índices gerados, é atribuido NaN
iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan

iris_2d[:10]


array([[b'5.1', b'3.5', b'1.4', b'0.2', b'Iris-setosa'],
       [b'4.9', b'3.0', b'1.4', b'0.2', b'Iris-setosa'],
       [b'4.7', b'3.2', b'1.3', b'0.2', b'Iris-setosa'],
       [b'4.6', b'3.1', b'1.5', b'0.2', b'Iris-setosa'],
       [b'5.0', b'3.6', b'1.4', b'0.2', b'Iris-setosa'],
       [b'5.4', b'3.9', b'1.7', b'0.4', b'Iris-setosa'],
       [b'4.6', b'3.4', b'1.4', b'0.3', b'Iris-setosa'],
       [b'5.0', b'3.4', b'1.5', b'0.2', b'Iris-setosa'],
       [b'4.4', nan, b'1.4', b'0.2', b'Iris-setosa'],
       [b'4.9', b'3.1', b'1.5', b'0.1', b'Iris-setosa']], dtype=object)

##### 33. How to find the position of missing values in numpy array?
    Difficulty Level: L2

    Q. Find the number and position of missing values in iris_2d's sepallength (1st column)

    # Input
    url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
    iris_2d = np.genfromtxt(url, delimiter=',', dtype='float')
    iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan

In [89]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris_2d = np.genfromtxt(url, delimiter=',', dtype='float')
iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan

# Para localizar valores NaN em arrays, é preciso usar a função np.isnan()
print(np.isnan(iris_2d[:, 0]).sum())
print(np.where(np.isnan(iris_2d[:,0])))

5
(array([ 38,  80, 106, 113, 121]),)


##### 34. How to filter a numpy array based on two or more conditions?
    Difficulty Level: L3

    Q. Filter the rows of iris_2d that has petallength (3rd column) > 1.5 and sepallength (1st column) < 5.0

In [90]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_2d = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0, 1, 2, 3])

# filtra os índices das linhas
mask = np.where((iris_2d[:, 2] > 1.5) & (iris_2d[:, 0] < 5))

# printa as linhas
iris_2d[mask]

array([[4.8, 3.4, 1.6, 0.2],
       [4.8, 3.4, 1.9, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [4.9, 2.4, 3.3, 1. ],
       [4.9, 2.5, 4.5, 1.7]])

##### 35. How to drop rows that contain a missing value from a numpy array?
    Difficulty Level: L3:

    Q. Select the rows of iris_2d that does not have any nan value.

In [91]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris_2d = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0, 1, 2, 3])

iris_2d[np.sum(np.isnan(iris_2d), axis=1) == 0][:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

##### 36. How to find the correlation between two columns of a numpy array?
    Difficulty Level: L2

    Q. Find the correlation between SepalLength(1st column) and PetalLength(3rd column) in iris_2d

In [92]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_2d = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0, 1, 2, 3])

sepallength = iris_2d[:, 0]
petallength = iris_2d[:, 2]

# a função np.corrcoef acha o índice d ecorrelação entre dois arrays de dimensão 1D
np.corrcoef(sepallength, petallength)[0, 1]

0.8717541573048718

##### 37. How to find if a given array has any null values?
    Difficulty Level: L2

    Q. Find out if iris_2d has any missing values.

In [93]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_2d = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0, 1, 2, 3])

# Retorna o número de NaN no array, que é dado como 0
np.isnan(iris_2d).sum()

0

##### 38. How to replace all missing values with 0 in a numpy array?
    Difficulty Level: L2

    Q. Replace all ccurrences of nan with 0 in numpy array

In [94]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_2d = np.genfromtxt(url, delimiter=",", dtype="float", usecols=[0, 1, 2, 3])
iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan

# Salva em uma máscara booleana as amostras com valor NaN
mask = np.isnan(iris_2d)

iris_2d[mask] = 0

iris_2d[:4]

array([[0. , 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2]])

##### 39. How to find the count of unique values in a numpy array?
    Difficulty Level: L2

    Q. Find the unique values and the count of unique values in iris's species

In [95]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris = np.genfromtxt(url, delimiter=',', dtype='object')
names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

species = iris[:, 4]

# np.unique retorna todos os valores distintos e a qtd de cada um
np.unique(species, return_counts=True)

(array([b'Iris-setosa', b'Iris-versicolor', b'Iris-virginica'],
       dtype=object),
 array([50, 50, 50]))

##### 40. How to convert a numeric to a categorical (text) array?
    Difficulty Level: L2

    Q. Bin the petal length (3rd) column of iris_2d to form a text array, such that if petal length is:

    Less than 3 --> 'small'
    3-5 --> 'medium'
    '>=5 --> 'large'

* Tentar aplicar no trabalho o digitize

In [96]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris = np.genfromtxt(url, delimiter=",", dtype="object")
names = ("sepallength", "sepalwidth", "petallength", "petalwidth", "species")

# Convertendo a terceira coluna para tipo float
# Além disso define os intervalos (bins) 0-3, 3-5, 5-10, fornecidos em formato de array
petal_length_bin = np.digitize(iris[:, 2].astype("float"), [0, 3, 5, 10])

label_map = {1: "small", 2: "medium", 3: "large", 4: np.nan}
petal_length_cat = [label_map[x] for x in petal_length_bin]

petal_length_cat[:4]

['small', 'small', 'small', 'small']

##### 41. How to create a new column from existing columns of a numpy array?
    Difficulty Level: L2

    Q. Create a new column for volume in iris_2d, where volume is (pi x petallength x sepal_length^2)/3

In [97]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris_2d = np.genfromtxt(url, delimiter=",", dtype="object")
names = ("sepallength", "sepalwidth", "petallength", "petalwidth", "species")

# ao atribuir uma coluna específica para uma variável, é possível definir o tipo dela, até mesmo altera-lo
sepallength = iris_2d[:, 0].astype('float')
petallength = iris_2d[:, 2].astype('float')

volume = (np.pi * petallength * (sepallength ** 2))/3

# muda a dimensão para que tenha uma única coluna
volume = np.reshape(volume, (-1,1))

iris_2d = np.hstack([iris_2d, volume])

iris_2d[0:5]

array([[b'5.1', b'3.5', b'1.4', b'0.2', b'Iris-setosa',
        38.13265162927291],
       [b'4.9', b'3.0', b'1.4', b'0.2', b'Iris-setosa',
        35.200498485922445],
       [b'4.7', b'3.2', b'1.3', b'0.2', b'Iris-setosa', 30.0723720777127],
       [b'4.6', b'3.1', b'1.5', b'0.2', b'Iris-setosa',
        33.238050274980004],
       [b'5.0', b'3.6', b'1.4', b'0.2', b'Iris-setosa',
        36.65191429188092]], dtype=object)

##### 42. How to do probabilistic sampling in numpy?
    Difficulty Level: L3

    Q. Randomly sample iris's species such that setose is twice the number of versicolor and virginica

In [98]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris = np.genfromtxt(url, delimiter=",", dtype="object")

species = iris[:, 4]

a = np.unique(species)

np.random.seed(100)
#np.random.choice escolhe aleatoriamente elementos do vetor a, passado para a função
# a escolha é baseada na % de chance atribuido a cada amostra, sendo 50% para a primeira e 25% para as demais
species_out = np.random.choice(a, 150, p=[0.5, 0.25, 0.25])

species_out

array([b'Iris-versicolor', b'Iris-setosa', b'Iris-setosa',
       b'Iris-virginica', b'Iris-setosa', b'Iris-setosa',
       b'Iris-versicolor', b'Iris-virginica', b'Iris-setosa',
       b'Iris-versicolor', b'Iris-virginica', b'Iris-setosa',
       b'Iris-setosa', b'Iris-setosa', b'Iris-setosa', b'Iris-virginica',
       b'Iris-virginica', b'Iris-setosa', b'Iris-virginica',
       b'Iris-setosa', b'Iris-setosa', b'Iris-virginica',
       b'Iris-virginica', b'Iris-setosa', b'Iris-setosa', b'Iris-setosa',
       b'Iris-setosa', b'Iris-setosa', b'Iris-virginica', b'Iris-setosa',
       b'Iris-versicolor', b'Iris-versicolor', b'Iris-setosa',
       b'Iris-setosa', b'Iris-setosa', b'Iris-virginica',
       b'Iris-virginica', b'Iris-setosa', b'Iris-virginica',
       b'Iris-versicolor', b'Iris-versicolor', b'Iris-versicolor',
       b'Iris-versicolor', b'Iris-setosa', b'Iris-setosa',
       b'Iris-versicolor', b'Iris-virginica', b'Iris-setosa',
       b'Iris-setosa', b'Iris-virginica', b'Iris

##### 43. How to get the second largest value of an array when grouped by another array?
    Difficulty Level: L2

    Q. What is the value of second longest petallength of species setosa

In [99]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris = np.genfromtxt(url, delimiter=',', dtype='object')
names = ("sepallength", "sepalwidth", "petallength", "petalwidth", "species")

# Seleciona apenas os indices de linha Iris-setosa
rows = np.where(iris[:,4] == b'Iris-setosa')

# Depois pega os valores de petalas os convertendo para float
petal_len_setosa = iris[rows, 2].astype('float')

# ordena os valores unicos
petal_len_setosa = np.unique(np.sort(petal_len_setosa))

# Seleciona o penultimo
petal_len_setosa[-2]

1.7

##### 44. How to sort a 2D array by a column
    Difficulty Level: L2

    Q. Sort the iris dataset based on sepallength column.

In [100]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris = np.genfromtxt(url, delimiter=",", dtype="object")
names = ("sepallength", "sepalwidth", "petallength", "petalwidth", "species")

# Ordena pela coluna 0
print(iris[iris[:, 0].argsort()][:20])

[[b'4.3' b'3.0' b'1.1' b'0.1' b'Iris-setosa']
 [b'4.4' b'3.2' b'1.3' b'0.2' b'Iris-setosa']
 [b'4.4' b'3.0' b'1.3' b'0.2' b'Iris-setosa']
 [b'4.4' b'2.9' b'1.4' b'0.2' b'Iris-setosa']
 [b'4.5' b'2.3' b'1.3' b'0.3' b'Iris-setosa']
 [b'4.6' b'3.6' b'1.0' b'0.2' b'Iris-setosa']
 [b'4.6' b'3.1' b'1.5' b'0.2' b'Iris-setosa']
 [b'4.6' b'3.4' b'1.4' b'0.3' b'Iris-setosa']
 [b'4.6' b'3.2' b'1.4' b'0.2' b'Iris-setosa']
 [b'4.7' b'3.2' b'1.3' b'0.2' b'Iris-setosa']
 [b'4.7' b'3.2' b'1.6' b'0.2' b'Iris-setosa']
 [b'4.8' b'3.0' b'1.4' b'0.1' b'Iris-setosa']
 [b'4.8' b'3.0' b'1.4' b'0.3' b'Iris-setosa']
 [b'4.8' b'3.4' b'1.9' b'0.2' b'Iris-setosa']
 [b'4.8' b'3.4' b'1.6' b'0.2' b'Iris-setosa']
 [b'4.8' b'3.1' b'1.6' b'0.2' b'Iris-setosa']
 [b'4.9' b'2.4' b'3.3' b'1.0' b'Iris-versicolor']
 [b'4.9' b'2.5' b'4.5' b'1.7' b'Iris-virginica']
 [b'4.9' b'3.1' b'1.5' b'0.1' b'Iris-setosa']
 [b'4.9' b'3.1' b'1.5' b'0.1' b'Iris-setosa']]


##### 45. How to find the most frequent value in a numpy array?
    Difficulty Level: L1

    Q. Find the most frequent value of petal length (3rd column) in iris dataset.

In [101]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris = np.genfromtxt(url, delimiter=",", dtype="object")

# retorna os valores distintos da coluna petallength e suas respectivas frequencias
# Dessa maneira, o índice de cada valor único está associado ao indice de sua frequencia
vals, counts = np.unique(iris[:, 2], return_counts=True)

# Assim, ao usar argmax(count) retorna o indice da maior contagem, que esta associada ao valor unico de maior frequencia
print(vals[np.argmax(counts)])

b'1.5'


##### 46. How to find the position of the first occurrence of a value greater than a given value?
    Difficulty Level: L2

    Q. Find the position of the first occurrence of a value greater than 1.0 in petalwidth 4th column of iris dataset.

In [102]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
iris = np.genfromtxt(url, delimiter=",", dtype="object")

# np.argwhere retorna os índices que cumprem a condição passada a ser aplicada em um array
np.argwhere(iris[:, 3].astype(float) > 1.0)[0]

array([50])

##### 47. How to replace all values greater than a given value to a given cutoff?
    Difficulty Level: L2

    Q. From the array a, replace all values greater than 30 to 30 and less than 10 to 10.

In [103]:
np.random.seed(100)
a = np.random.uniform(1, 50, 20)

values_greater_30 = np.argwhere(a > 30)
values_less_10 = np.argwhere(a < 10)

a[values_greater_30] = 30
a[values_less_10] = 10

a

array([27.626842, 14.6401  , 21.801362, 30.      , 10.      , 10.      ,
       30.      , 30.      , 10.      , 29.179573, 30.      , 11.250904,
       10.081083, 10.      , 11.765177, 30.      , 30.      , 10.      ,
       30.      , 14.429614])

##### 48. How to get the positions of top n values from a numpy array?
    Difficulty Level: L2

    Q. Get the positions of top 5 maximum values in a given array a.

In [104]:
np.random.seed(100)
a = np.random.uniform(1, 50, 20)

# Retorna as posições
print(a.argsort())

# Retorna os 5 maiores valores
np.sort(a)[-1:-6:-1]

[ 4 13  5  8 17 12 11 14 19  1  2  0  9  6 16 18  7  3 10 15]


array([48.952565, 44.674776, 42.39403 , 41.466785, 40.995013])

##### 49. How to compute the row wise counts of all possible values in an array?
    Difficulty Level: L4

    Q. Compute the counts of unique values row-wise.

In [105]:
np.random.seed(100)
arr = np.random.randint(1, 11, size=(6, 10))


def counts_of_all_values_rowwise(arr2d):
    num_counts_array = [np.unique(row, return_counts=True) for row in arr2d]

    return [
        [int(b[a == i]) if i in a else 0 for i in np.unique(arr2d)]
        for a, b in num_counts_array
    ]


# Print
print(np.arange(1, 11))
counts_of_all_values_rowwise(arr)

[ 1  2  3  4  5  6  7  8  9 10]


/tmp/ipykernel_27882/4044348034.py:9: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  [int(b[a == i]) if i in a else 0 for i in np.unique(arr2d)]


[[1, 0, 2, 1, 1, 1, 0, 2, 2, 0],
 [2, 1, 3, 0, 1, 0, 1, 0, 1, 1],
 [0, 3, 0, 2, 3, 1, 0, 1, 0, 0],
 [1, 0, 2, 1, 0, 1, 0, 2, 1, 2],
 [2, 2, 2, 0, 0, 1, 1, 1, 1, 0],
 [1, 1, 1, 1, 1, 2, 0, 0, 2, 1]]

##### 50. How to convert an array of arrays into a flat 1d array?
  Difficulty Level: 2

  Q. Convert array_of_arrays into a flat linear 1d array.

In [106]:
arr1 = np.arange(0, 3)
arr2 = np.arange(3, 6)
arr3 = np.arange(6, 9)
array_of_arrays = np.array([arr1, arr2, arr3])

result = np.hstack([array_of_arrays[0], array_of_arrays[1], array_of_arrays[2]])

result

array([0, 1, 2, 3, 4, 5, 6, 7, 8])

##### 51. How to generate one-hot encodings for an array in numpy?
  Difficulty Level L4

  Q. Compute the one-hot encodings (dummy binary variables for each unique value in the array)

In [107]:
np.random.seed(101)
arr = np.random.randint(1, 4, size=6)


def one_hot_encodings(arr):
    uniqs = np.unique(arr)
    out = np.zeros((arr.shape[0], uniqs.shape[0]))
    for i, k in enumerate(arr):
        out[i, k - 1] = 1
    return out


one_hot_encodings(arr)

array([[0., 1., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.],
       [1., 0., 0.]])

##### 52. How to create row numbers grouped by a categorical variable?
  Difficulty Level: L3

  Q. Create row numbers grouped by a categorical variable. Use the following sample from iris species as input.

In [108]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
species = np.genfromtxt(url, delimiter=",", dtype="str", usecols=4)
species_small = np.sort(np.random.choice(species, size=20))


print(
    [
        i
        for val in np.unique(species_small)
        for i, grp in enumerate(species_small[species_small == val])
    ]
)

[0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 6, 7, 8]


#### 53. How to create groud ids based on a given categorical variable?

    Difficulty Level: L4

    Q. Create group ids based on a given categorical variable. Use the following sample from iris species as input.

In [111]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
species = np.genfromtxt(url, delimiter=',', dtype='str', usecols=4)
species_small = np.sort(np.random.choice(species, size=20))
species_small

output = [np.argwhere(np.unique(species_small) == s).tolist()[0][0] for val in np.unique(species_small) for s in species_small[species_small==val]]

output

[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2]

##### 54. How to rank items in an array using numpy?

Difficulty Level: L2

Q. Create the ranks for the given numeric array a.

In [116]:
np.random.seed(10)
a = np.random.randint(20, size=10)
# argsort fornece o índice que ordenariam um array, ou seja, a respectiva posição que cada elemento deveria tomar
a.argsort().argsort()

array([4, 2, 6, 0, 8, 7, 9, 3, 5, 1])

#### 55. How to rank items in a multidimensional array using numpy?

Difficulty Level: L3

Q. Create a rank array of the same shape as a given numeric array a.

In [119]:
np.random.seed(10)
a = np.random.randint(20, size=[2,5])

print(a.ravel().argsort().argsort().reshape(a.shape))

array([[ 9,  4, 15,  0, 17],
       [16, 17,  8,  9,  0]])

#### 56. How to find the maximum value in each row of a numpy array 2d?

DifficultyLevel: L2

Q. Compute the maximum for each row in the given array.

In [121]:
np.random.seed(100)
a = np.random.randint(1,10, [5,3])

# recupera o máximo em cada eixo 1, linhas
np.max(a, axis=1)

array([9, 8, 6, 3, 9])

#### 57. How to compute the min-by-max for each row for a numpy array 2d?

DifficultyLevel: L3

Q. Compute the min-by-max for each row for given 2d numpy array.

In [123]:
np.random.seed(100)
a = np.random.randint(1,10, [5,3])

np.apply_along_axis(lambda x: np.min(x)/np.max(x), arr=a, axis=1)

array([0.444444, 0.125   , 0.5     , 1.      , 0.111111])

#### 58. How to find the duplicate records in a numpy array?

Difficulty Level: L3

Q. Find the duplicate entries (2nd occurrence onwards) in the given numpy array and mark them as True. First time occurrences should be False.

In [125]:
np.random.seed(100)
a = np.random.randint(0, 5, 10)

arr = np.full_like(a.shape[0], True)
unique_positions =np.unique(a, return_index=True)[1]
out[unique_positions] = False

print(out)

[ 0 -1  0 -1  0  0  6 -1  8 -1]


##### 59. How to find the grouped mean in numpy?

Difficulty Level L3

Q. Find the mean of a numeric column grouped by a categorical column in a 2D numpy array

In [126]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris = np.genfromtxt(url, delimiter=',', dtype='object')
names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

numeric_column = iris[:, 1].astype('float')  
grouping_column = iris[:, 4]  


#### 60. How to convert a PIL image to numpy array?

Difficulty Level: L3

Q. Import the image from the following URL and convert it to a numpy array.

URL = 'https://upload.wikimedia.org/wikipedia/commons/8/8b/Denali_Mt_McKinley.jpg'

In [132]:
from io import BytesIO
from PIL import Image
import PIL, requests

# Import image from URL
URL = 'https://static.mundoeducacao.uol.com.br/mundoeducacao/2019/09/coracao.jpg'
response = requests.get(URL)

# Read it as Image
I = Image.open(BytesIO(response.content))

# Optionally resize
I = I.resize([150,150])

# Convert to numpy array
arr = np.asarray(I)


#### 61. How to drop all missing values from a numpy array?

Difficulty Level: L2

Q. Drop all nan values from a 1D numpy array

In [138]:
arr = np.array([1,2,3,np.nan,5,6,7,np.nan])

# aplicação de uma máscara
# verifica quais posições são iguals a np.nan
# np.isna retorna true para todos as posições que forem NaN
mask = np.isnan(arr)

# seleciona todos que NÃO SÃO np.nan
arr[~mask]

array([False, False, False,  True, False, False, False,  True])

#### 62. How to compute the euclidean distance between two arrays?

Difficulty Level: L3

Q. Compute the euclidean distance between two arrays a and b.

In [140]:
a = np.array([1,2,3,4,5])
b = np.array([4,5,6,7,8])

# np.linalg calcula a norma de um vetor, no caso abaixo, está normalizando a - b
dist = np.linalg.norm(a-b)
dist

6.708203932499369

#### 63. How to find all the local maxima (or peaks) in a 1d array?

Difficulty Level: L4

Q. Find all the peaks in a 1D numpy array a. Peaks are points surrounded by smaller values on both sides.

In [141]:
a = np.array([1, 3, 7, 1, 2, 6, 0, 1])
doublediff = np.diff(np.sign(np.diff(a)))
peak_locations = np.where(doublediff == -2)[0] + 1
peak_locations

array([2, 5])

##### 64. How to subtract a 1d array from a 2d array, where each item of 1d array subtracts from respective row?

Difficulty Level: L2

Q. Subtract the 1d array b_1d from the 2d array a_2d, such that each item of b_1d subtracts from respective row of a_2d.

In [148]:
a_2d = np.array([[3,3,3],[4,4,4],[5,5,5]])
b_1d = np.array([1,1,1])

arr =np.subtract(a_2d, b_1d[:,None])

arr

[[2 2 2]
 [3 3 3]
 [4 4 4]]


##### 65. How to find the index of n'th repetition of an item in an array

Difficulty Level L2

Q. Find the index of 5th repetition of number 1 in x.

In [149]:
x = np.array([1, 2, 1, 1, 3, 4, 3, 1, 1, 2, 1, 1, 2])
n = 5

[i for i, v in enumerate(x) if v == 1][n-1]

8

##### 66. How to convert numpy's datetime64 object to datetime's datetime object?

Difficulty Level: L2

Q. Convert numpy's datetime64 object to datetime's datetime object

In [150]:
from datetime import datetime
dt64 = np.datetime64('2018-02-25 22:10:10')

# astype converte para o tipo passado no parâmetro
dt64.astype(datetime)

datetime.datetime(2018, 2, 25, 22, 10, 10)

##### 67. How to compute the moving average of a numpy array?

Difficulty Level: L3

Q. Compute the moving average of window size 3, for the given 1D array.

In [151]:
def moving_average(a, n=3) :
    ret = np.cumsum(a, dtype=float)
    ret[n:] = ret[n:] - ret[:-n]
    return ret[n - 1:] / n

np.random.seed(100)
Z = np.random.randint(10, size=10)
print('array: ', Z)
moving_average(Z, n=3).round(2)

array:  [8 8 3 7 7 0 4 2 5 2]


array([6.33, 6.  , 5.67, 4.67, 3.67, 2.  , 3.67, 3.  ])

##### 68. How to create a numpy array sequence given only the starting point, length and the step?

Difficulty Level: L2

Q. Create a numpy array of length 10, starting from 5 and has a step of 3 between consecutive numbers

In [152]:
length = 10
start = 5
step = 3

def seq(start, length, step):
    end = start + (step*length)
    return np.arange(start, end, step)

seq(start, length, step)

array([ 5,  8, 11, 14, 17, 20, 23, 26, 29, 32])

##### 69. How to fill in missing dates in an irregular series of numpy dates?

Difficulty Level: L3

Q. Given an array of a non-continuous sequence of dates. Make it a continuous sequence of dates, by filling in the missing dates.

In [153]:
dates = np.arange(np.datetime64('2018-02-01'), np.datetime64('2018-02-25'), 2)
print(dates)

filled_in = np.array([np.arange(date, (date+d)) for date, d in zip(dates, np.diff(dates))]).reshape(-1)


['2018-02-01' '2018-02-03' '2018-02-05' '2018-02-07' '2018-02-09'
 '2018-02-11' '2018-02-13' '2018-02-15' '2018-02-17' '2018-02-19'
 '2018-02-21' '2018-02-23']


##### 70. How to create strides from a given 1D array?

Difficulty Level: L4

Q. From the given 1d array arr, generate a 2d matrix using strides, with a window length of 4 and strides of 2, like [[0,1,2,3], [2,3,4,5], [4,5,6,7]..] 

In [154]:
def gen_strides(a, stride_len=5, window_len=5):
    n_strides = ((a.size-window_len)//stride_len) + 1
    return np.array([a[s:(s+window_len)] for s in np.arange(0, n_strides*stride_len, stride_len)])

print(gen_strides(np.arange(15), stride_len=2, window_len=4))


[[ 0  1  2  3]
 [ 2  3  4  5]
 [ 4  5  6  7]
 [ 6  7  8  9]
 [ 8  9 10 11]
 [10 11 12 13]]
